packages/sources/token-balance/src/transport/evm.ts

```diff
@@ -0,0 +1,244 @@
+      `${network}_${contractAddress}`
+
+    const decimalsMap = new Map<string, number | null>()
+    const decimalsRequests = []
+    for (const address of normalizedAddresses) {
+      const { network, contractAddress, decimalsSignature, provider } = address
+
+      // since we can have multiple addresses with balances of the same token,
+      // skip if we've encountered this pair already
+      const decimalsKey = constructDecimalsKey(network, contractAddress)
+      if (decimalsMap.has(decimalsKey)) {
+        logger.debug(`skipping decimals fetch for contract address ${address}`)
+        continue
+      }
+      decimalsMap.set(decimalsKey, null)
+
+      // get decimals
+      const iface = new ethers.Interface([decimalsSignature])
+      const decimalsFunctionName = iface.getFunctionName(decimalsSignature)
+      const decimalsEncoded = iface.encodeFunctionData(decimalsFunctionName)
+
+      // push with callback including network and contract address
+      decimalsRequests.push(
+        provider
+          .call({
+            to: contractAddress,
+            data: decimalsEncoded,
+          })
+          .then((result) => ({
+            network,
+            contractAddress,
+            decimals: Number(iface.decodeFunctionResult(decimalsSignature, result)[0]),
+          })),
+      )
+    }
+
+    const decimalsResponses = await Promise.all(decimalsRequests)
+
+    decimalsResponses.forEach((response) => {
+      const decimalsKey = constructDecimalsKey(response.network, response.contractAddress)
+      decimalsMap.set(decimalsKey, response.decimals)
+    })
+
+    // construct list of RPC balance requests
+    const balanceRequests = []
+    for (const address of normalizedAddresses) {
+      const { network, contractAddress, wallets, balanceOfSignature, provider } = address
+      const iface = new ethers.Interface([balanceOfSignature])
+      const balanceOfFunctionName = iface.getFunctionName(balanceOfSignature)
+
+      for (const wallet of wallets) {
+        const balanceOfEncoded = iface.encodeFunctionData(balanceOfFunctionName, [wallet])
+        balanceRequests.push(
+          provider
+            .call({
+              to: contractAddress,
+              data: balanceOfEncoded,
+            })
+            .then((result) => ({
+              network: network.toLowerCase(), // return lowercase
+              contractAddress,
+              walletAddress: wallet,
+              balance: String(iface.decodeFunctionResult(balanceOfSignature, result)[0]),
+              decimals: decimalsMap.get(constructDecimalsKey(network, contractAddress)) || 0,
+            })),
+        )
+      }
+    }
+
+    const balanceResponses = await Promise.all(balanceRequests)
+
+    // compute result by scaling all to 18 decimals, handles
+    const result = balanceResponses.reduce(
+      (accumulator, current) =>
+        accumulator + Number(current.balance) * Math.pow(10, 18 - current.decimals),
```

Comment:
You might need bigint here

Reply:
I was doing it with ethers.toBigInt but
1. this doesn't allow for decimals > 18 eg: https://etherscan.io/token/0xaba8cac6866b83ae4eec97dd07ed254282f6ad8a#readContract
2. couldn't get ethers.FixedNumber to work for > 18
3. `Number.MAX_VALUE = 1.7976931348623157e+308`

Reply:
I guess theoretically we could still overflow with an even weirder token, but e+308 is pretty high.

Reply:
does bigint(current.balance) work?
Also I think limit is really Number.MAX_SAFE_INTEGER

Reply:
I think both sides of the mult have to be bigints

Reply:
bigint(accumulator) + bigint(current.balance) * bigint(Math.pow(10, 18 - current.decimals))

Reply:
the other option is we can just set 18 decimals as max (+ use bigint) and error otherwise for now

Reply:
I see you can do a if else then? > 18 we divide < 18 we *

Reply:
yeah true, ok will do something like `floor(value / Math.pow(10, decimals - 18))`